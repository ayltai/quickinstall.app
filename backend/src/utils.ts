import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
