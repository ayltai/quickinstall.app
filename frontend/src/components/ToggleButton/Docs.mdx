import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './ToggleButton.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use toggle buttons

Use a toggle button to let users quickly switch the selection of an item.

When using toggle buttons,
* Once a user selects an item, the results should be displayed immediately
* Provide a visible label that clearly identifies the purpose of the toggle buttons

# Guidelines

* Be clear about what will happen if the toggle is selected
* Start your toggle labels with a capital letter
* Don't use commas or any punctuation at tht end of each label
* Don't use more than two words per label

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Examples'
    includePrimary={false} />
