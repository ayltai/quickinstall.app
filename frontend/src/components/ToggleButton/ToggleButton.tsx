import React, { type FC, useState, } from 'react';

import { MaterialButton, } from '../Button';
import type { ToggleButtonProps, } from './ToggleButton.types';

/**
 * A toggle button lets users switch between the selection of an item on or off.
 * @param icon The icon to show when the button is in the unselected state
 * @param selectedIcon The icon to show when the button is in the selected state
 * @param deselectIcon The icon to show when the button is in the selected state and is hovered
 * @param text The text to show when the button is in the unselected state
 * @param selectedText The text to show when the button is in the selected state
 * @param deselectText The text to show when the button is in the selected state and is hovered
 * @param onChange Function to call when the selected state is changed
 * @param props Other props
 */
export const ToggleButton : FC<ToggleButtonProps> = ({
    icon,
    selectedIcon,
    deselectIcon,
    text,
    selectedText,
    deselectText,
    onChange,
    ...props
}) => {
    const { color, selected, ...rest } = props;

    const [ hover, setHover, ] = useState(false);

    const handleClick = () => onChange && onChange(!selected);

    const handleMouseOver = () => setHover(true);

    const handleMouseOut = () => setHover(false);

    return (
        <MaterialButton
            aria-checked={selected}
            role='checkbox'
            color={color}
            variant={selected ? 'contained' : 'text'}
            startIcon={selected ? hover ? deselectIcon : selectedIcon : icon}
            onClick={handleClick}
            onMouseOver={handleMouseOver}
            onMouseOut={handleMouseOut}
            {...rest}>
            {selected ? hover ? deselectText : selectedText : text}
        </MaterialButton>
    );
};
