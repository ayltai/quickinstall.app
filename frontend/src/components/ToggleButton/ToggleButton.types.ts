import type { ToggleButtonOwnProps, } from '@mui/material/ToggleButton';
import type { ReactNode, } from 'react';

export type ToggleButtonProps = Omit<ToggleButtonOwnProps, 'color' | 'value' | 'onClick' | 'onChange'> & {
    color?        : 'inherit' | 'primary' | 'secondary' | 'success' | 'error' | 'info' | 'warning',
    icon?         : ReactNode,
    selectedIcon? : ReactNode,
    deselectIcon? : ReactNode,
    text?         : ReactNode,
    selectedText? : ReactNode,
    deselectText? : ReactNode,
    onClick?      : () => void,
    onChange?     : (selected : boolean) => void,
};
