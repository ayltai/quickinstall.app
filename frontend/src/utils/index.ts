import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
