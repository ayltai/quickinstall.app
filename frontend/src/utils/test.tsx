import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
