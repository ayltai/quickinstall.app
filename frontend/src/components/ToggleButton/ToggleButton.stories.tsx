import { Check, DeleteForever, } from '@mui/icons-material';
import type { Meta, StoryObj, } from '@storybook/react';
import React, { useState, } from 'react';

import { ToggleButton, } from './ToggleButton';

const meta : Meta<typeof ToggleButton> = {
    component : ToggleButton,
    title     : 'Components/Actions/Toggle',
    argTypes  : {
        icon          : {
            description : 'The icon to show when the button is in the unselected state',
        },
        selectedIcon  : {
            description : 'The icon to show when the button is in the selected state',
        },
        deselectIcon  : {
            description : 'The icon to show when the button is in the selected state and is hovered',
        },
        text          : {
            description : 'The text to show when the button is in the unselected state',
        },
        selectedText  : {
            description : 'The text to show when the button is in the selected state',
        },
        deselectText  : {
            description : 'The text to show when the button is in the selected state and is hovered',
        },
        onChange      : {
            description : 'Function to call when the selected state is changed',
        },
    },
};

export default meta;

type Story = StoryObj<typeof ToggleButton>;

export const Default : Story = {
    args   : {
        icon          : <Check />,
        selectedIcon  : <Check />,
        deselectIcon  : <DeleteForever />,
        text          : 'Select',
        selectedText  : 'Selected',
        deselectText  : 'Remove',
    },
    render : props => {
        const [ selected, setSelected, ] = useState(props.selected ?? false);

        return (
            <ToggleButton
                {...props}
                selected={selected}
                onChange={setSelected} />
        );
    },
};

export const Selected : Story = {
    args : {
        ...Default.args,
        selected : true,
    },
};
