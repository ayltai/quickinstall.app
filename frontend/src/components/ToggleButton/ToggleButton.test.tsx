import { CheckCircle, DeleteForever, } from '@mui/icons-material';
import React from 'react';

import { fireEvent, render, } from '../../utils/test';
import { ToggleButton, } from './ToggleButton';

describe('<ToggleButton />', () => {
    it('renders default style correctly', () => expect(render(
        <ToggleButton
            icon={<CheckCircle />}
            selectedIcon={<CheckCircle />}
            deselectIcon={<DeleteForever />}
            text='Select'
            selectedText='Selected'
            deselectText='Remove' />
    )).toMatchSnapshot());

    it('renders selected style correctly', () => expect(render(
        <ToggleButton
            selected
            icon={<CheckCircle />}
            selectedIcon={<CheckCircle />}
            deselectIcon={<DeleteForever />}
            text='Select'
            selectedText='Selected'
            deselectText='Remove' />
    )).toMatchSnapshot());

    it('triggers onChange when clicked', () => {
        const handleChange = vi.fn();

        const { getByRole, } = render(
            <ToggleButton
                icon={<CheckCircle />}
                selectedIcon={<CheckCircle />}
                deselectIcon={<DeleteForever />}
                text='Select'
                selectedText='Selected'
                deselectText='Remove'
                onChange={handleChange} />
        );

        fireEvent.click(getByRole('checkbox'));

        expect(handleChange).toBeCalledTimes(1);
        expect(handleChange).toBeCalledWith(true);
    });
});
