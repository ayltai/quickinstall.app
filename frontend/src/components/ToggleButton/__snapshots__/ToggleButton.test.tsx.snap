// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ToggleButton /> > renders default style correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-lzklvy-MuiButtonBase-root-MuiButton-root"
        role="checkbox"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
            />
          </svg>
        </span>
        Select
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-lzklvy-MuiButtonBase-root-MuiButton-root"
      role="checkbox"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
          />
        </svg>
      </span>
      Select
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<ToggleButton /> > renders selected style correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-checked="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1d0nnqb-MuiButtonBase-root-MuiButton-root"
        role="checkbox"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
            />
          </svg>
        </span>
        Selected
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-checked="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1d0nnqb-MuiButtonBase-root-MuiButton-root"
      role="checkbox"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
          />
        </svg>
      </span>
      Selected
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
